R/benchTradePerf.R

```diff
@@ -0,0 +1,214 @@
+              benchPrice <- vwapPrice
+              
+              out <- as.data.frame(cbind(Symbol, c("Buy", "Sell")[side], p_avg, benchPrice), stringsAsFactors = FALSE)
+              colnames(out) <- c('Symbol', 'Side', 'Avg.Exec.Price', paste(benchmark[i], type, sep = '.'))
+            },
+            PWP = {
+              if (is.null(POV)) stop(paste("POV rate needed to compute", benchmark))
+              # if (!("MktVolmn" %in% colnames(MktData))) stop("No MktVolmn column found, what did you call it?")
+              # if (missing(arrTime)) arrTime <- first(index(txns))
+              
+              pwpShares <- rep(tTxnQty/POV, length(txns$Txn.Price))
+              pwpPrice  <- crossprod(txns$Txn.Price, pwpShares)/sum(pwpShares)
+              
+              benchPrice <- pwpPrice
+              
+              out <- as.data.frame(cbind(Symbol, c("Buy", "Sell")[side], p_avg, POV, sum(pwpShares), benchPrice), stringsAsFactors = FALSE)
+              colnames(out) <- c('Symbol', 'Side', 'Avg.Exec.Price', 'POV', 'PWP.Shares', 'PWP.Price')
+            },
+            RPM = {
+              if (!(("MktPrice" %in% colnames(MktData)) & ("MktVolmn" %in% colnames(MktData)))) stop("No MktPrice or MktVolmn column found, what did you call them?")
+              
+              tMktVolmn <- sum(MktData[, "MktVolmn"])
+              
+              if (side == 1) {
+                tFavVolmn   <- sum(MktData[, "MktVolmn"][MktData[, "MktPrice"] > p_avg])
+                tUnfavVolmn <- sum(MktData[, "MktVolmn"][MktData[, "MktPrice"] < p_avg])
+              } else {
+                tFavVolmn   <- sum(MktData[, "MktVolmn"][MktData[, "MktPrice"] < p_avg])
+                tUnfavVolmn <- sum(MktData[, "MktVolmn"][MktData[, "MktPrice"] > p_avg])
+              }
+              
+              rpm <- 0.5*(tMktVolmn + tFavVolmn - tUnfavVolmn)/tMktVolmn
+              
+              out <- as.data.frame(cbind(Symbol, c("Buy", "Sell")[side], p_avg, tMktVolmn, tFavVolmn, tUnfavVolmn, rpm), stringsAsFactors = FALSE)
+              colnames(out) <- c('Symbol', 'Side', 'Avg.Exec.Price', 't.Mkt.Volmn', 't.Fav.Volmn', 't.Unfav.Volmn', benchmark[i])
+              
+              # Append a RPM qualitative score
+              if (verbose) { 
+                if (rpm >= 0.4 & rpm <= 0.6) {
+                  quality <- "Average"
+                } else if (rpm > 0.6) {
+                  quality <- "Good"
+                  if (rpm >= 0.8) {
+                    quality <- "Excellent" 
+                  }
+                } else {
+                  quality <- "Fair"
+                  if (rpm < 0.2) {
+                    quality <- "Poor"
+                  }
+                }
+                out[, 'Quality'] <- quality 
+              }
+            }
+    )
+    
+    # PnL performance for 'MktBench', 'VWAP' (all types) and 'PWP' 
+    if (benchmark[i] != 'RPM') {
+      out[, 'Performance'] <- (-1) * side * (p_avg - benchPrice)/benchPrice * 10000
```

Comment:
This really should not happen. Also, in the editing you have done on the original comment I have noticed that a new example with `type = 'Txns'` was introduced and that it shows a `Txn.Qty` column instead of `Performance`. But if you check the object, the name is correctly 'Performance' in both cases.

Reply:
i think i see the problem...benchPrice is a matrix result from crossprod()...if you use benchPrice[1,1] it works. Alternatively as opposed to using crossprod() just use cumsum*cumsum?

Reply:
Yes, the new example test with type='Txns' was to cover the other scenario

Reply:
Great solution. However, it is a curious error, `pwpPrice` goes the same way but it does not fail in giving the correct 'Performance' column. I found a different solution which produces corrects results and it is perhaps easier to maintain.

Reply:
Solved with f7eea99

Reply:
happy